### PR TITLE
gateway-api: attach routes to egress services

### PR DESCRIPTION
**Pull request by frankbu**
**Please provide a description of this PR:**

Allow Istio GAMMA users to set a route `parentRef` to a `ServiceEntry` to configure mesh traffic for external services. This allows users to do things like add headers, weighted routing, etc., for services outside of the mesh. Most importantly, this will allow users to redirect external traffic to an egress gateway using Gateway API `Gateway` and `xxxRoute` configuration resources, instead of Istio `Gateway` and `VirtualService`: https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway/

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/testdata/serviceentry.yaml [hunk 1]
```diff
@@ -23,8 +23,37 @@ spec:
...
   rules:
-  -
-    backendRefs:
+  - backendRefs:
+    - kind: Hostname
+      group: networking.istio.io
+      name: google.com
+      port: 80
+---
+apiVersion: networking.istio.io/v1alpha3
```
Comment (ymesika) [inline pilot/pkg/config/kube/gateway/testdata/serviceentry.yaml:32]:
```suggestion
apiVersion: networking.istio.io/v1beta1
```

**Comment by keithmattix** [comment]
@frankbu could you add the use-case and overview of the implementation in the description to aid in review?

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 2]
```diff
@@ -820,11 +842,13 @@ func toInternalParentReference(p k8s.ParentReference, localNamespace string) (pa
 	group := ptr.OrDefault((*string)(p.Group), gvk.KubernetesGateway.Group)
 	var ik config.GroupVersionKind
 	var ns string
-	// Currently supported types are Gateway and Service
+	// Currently supported types are Gateway, Service, and ServiceEntry
 	if kind == gvk.KubernetesGateway.Kind && group == gvk.KubernetesGateway.Group {
 		ik = gvk.KubernetesGateway
 	} else if kind == gvk.Service.Kind && group == gvk.Service.Group {
 		ik = gvk.Service
+	} else if kind == gvk.ServiceEntry.Kind && group == gvk.ServiceEntry.Group {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/conversion.go:850]: I thought we have kind=="Hostname"?

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:850]: Oh nvm, that is the backend not the frontend

**Review by howardjohn — APPROVED** [review]

**Review by keithmattix — APPROVED** [review]

**Pull request merged**